lib/dense.ml

```diff
@@ -239,6 +239,15 @@ let mapi_by_col ?(d=0) f x =
 
 let map_by_col ?(d=0) f x = mapi_by_col ~d (fun _ y -> f y) x
 
+let map2i f x y =
+  if row_num x = row_num y && col_num x = col_num y then
+    let z = empty (row_num x) (col_num x) in
+    iteri (fun i j w -> Array2.unsafe_set z i j (f i j w)) x; y
```

Comment:
If I understand the meaning of `map2i` correctly, maybe the type of `f` should be `val int -> int -> float -> float -> float` here, as it should take the values from both matrix `x` and `y`. Thus here it should be like something as
```ocaml
iteri (fun i j vx ->
  let vy = Array2.unsafe_get y i j in
  Array2.unsafe_set z i j (f i j vx vy)) x;
z
```
And the following `map2` function could be:
```ocaml
let map2 f x y = map2i (fun _ _ vx vy -> f vx vy) x y
```